File: baekjoon/baekjoon_1463.java
// https://www.acmicpc.net/problem/1463
// 백준 1463 : 1로 만들기

import java.util.Scanner;

public class baekjoon_1463 {

    static int[] dp;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();

        dp = new int[1000001];
        dp[1] = 0;

        for (int i = 2; i <= N; i++) {
            dp[i] = dp[i - 1] + 1;
            if (i % 2 == 0) {
                dp[i] = Math.min(dp[i], dp[i / 2] + 1);
            }
            if (i % 3 == 0) {
                dp[i] = Math.min(dp[i], dp[i / 3] + 1);
            }
        }

        System.out.println(dp[N]);
    }
}

File: baekjoon/baekjoon_1463.py
# https://www.acmicpc.net/problem/1463
# 백준 1463 : 1로 만들기
# LEVEL : Silver 3

import sys


def solution():
    N = int(sys.stdin.readline())

    dp = [N for _ in range(N+1)]
    dp[N] = 0

    for num in range(N, 0, -1):
        if dp[num] != N:
            if num % 3 == 0:
                dp[num//3] = min(dp[num//3], dp[num]+1)
            if num % 2 == 0:
                dp[num//2] = min(dp[num//2], dp[num]+1)
            dp[num-1] = min(dp[num-1], dp[num]+1)

    print(dp[1])


solution()
